File: framework/src/main/java/com/linxiao/framework/language/AppLanguageHelper.java
package com.linxiao.framework.language;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.util.DisplayMetrics;


import com.linxiao.framework.common.ContextProviderKt;
import com.linxiao.framework.preferences.AppPreferences;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * app language select helper class
 *
 * @author lx8421bcd
 * @since 2022-11-08
 */
public class AppLanguageHelper {

    public static final List<LanguageOption> SUPPORTED_LANGUAGES = new ArrayList<>();

    static {
        initSupportLanguages();
    }

    public static void initSupportLanguages() {
        SUPPORTED_LANGUAGES.clear();
        SUPPORTED_LANGUAGES.add(new LanguageOption(Locale.ENGLISH, "English"));
        SUPPORTED_LANGUAGES.add(new LanguageOption(Locale.TRADITIONAL_CHINESE, "简体中文"));
    }

    private static final String PREF_SELECTED_LOCALE = "SELECTED_LOCALE";

    private static LanguageOption currentLanguageOption = null;

    public static Locale getSystemCurrentLocale() {
        return Resources.getSystem().getConfiguration().locale;
    }

    public static LanguageOption getFollowingSystemOption() {
        return LanguageOption.followingSystem();
    }

    private synchronized static LanguageOption getCachedLanguageOption() {
        LanguageOption cachedLocale = AppPreferences.getDefault().getSerializable(PREF_SELECTED_LOCALE);
        if (cachedLocale != null) {
            for (LanguageOption locale : SUPPORTED_LANGUAGES) {
                if (locale.getId().equals(cachedLocale.getId())) {
                    return locale;
                }
            }
        }
        return getFollowingSystemOption();
    }

    public synchronized static LanguageOption getCurrentLanguageOption() {
        if (currentLanguageOption == null) {
            currentLanguageOption = getCachedLanguageOption();
        }
        return currentLanguageOption;
    }

    public synchronized static void setLanguageFollowingSystem() {
        setLanguage(LanguageOption.followingSystem());
    }

    public synchronized static void setLanguageWithBroadcast(LanguageOption option) {
        setLanguage(option);
        EventBus.getDefault().post(new LanguageChangedEvent());
    }

    public synchronized static void setLanguageWithRestart(LanguageOption option) {
        setLanguage(option);
        Single.just(0)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .doOnSuccess(o -> {
            android.os.Process.killProcess(android.os.Process.myPid());
            System.exit(0);
        })
        .subscribe();
    }

    public synchronized static void setLanguage(LanguageOption option) {
        AppPreferences.getDefault().put(PREF_SELECTED_LOCALE, option);
        currentLanguageOption = option;
        Resources res = ContextProviderKt.getGlobalContext().getResources();
        changeResourcesConfig(res);
    }

    public synchronized static void updateLanguageSetting(Context context) {
        changeResourcesConfig(context.getResources());
    }

    public synchronized static void doOnContextGetResources(Resources res) {
        changeResourcesConfig(res);
    }

    private synchronized static void changeResourcesConfig(Resources res) {
        DisplayMetrics dm = res.getDisplayMetrics();
        Configuration conf = res.getConfiguration();
        conf.locale = getCurrentLanguageOption().getLocale();
        res.updateConfiguration(conf, dm);
    }
}

File: framework/src/main/java/com/linxiao/framework/language/AppLanguageHelper.kt
package com.linxiao.framework.language

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.res.Resources
import android.os.Process
import androidx.core.content.ContextCompat
import com.linxiao.framework.common.globalContext
import com.linxiao.framework.language.LanguageOption.Companion.followingSystem
import com.linxiao.framework.preferences.AppPreferences
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.Locale


/**
 * app language select helper class
 *
 * @author lx8421bcd
 * @since 2022-11-08
 */
object AppLanguageHelper {

    private class LanguageChangeReceiver : BroadcastReceiver() {

        companion object {
            const val LANGUAGE_CHANGED_ACTION = "LANGUAGE_CHANGED_ACTION"
        }

        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == LANGUAGE_CHANGED_ACTION) {
                if (context is Activity) {
                    context.recreate()
                }
            }
        }

        fun bindActivity(activity: Activity) {
            val filter = IntentFilter()
            filter.addAction(LANGUAGE_CHANGED_ACTION)
            ContextCompat.registerReceiver(activity,this, filter, ContextCompat.RECEIVER_NOT_EXPORTED)
        }

        fun unbindActivity(activity: Activity) {
            activity.unregisterReceiver(this)
        }
    }

    private const val PREF_SELECTED_LOCALE = "SELECTED_LOCALE"

    private val subscribedBroadcastReceivers = HashMap<Int, LanguageChangeReceiver>()
    private val supportedLanguages: MutableList<LanguageOption> = ArrayList()
    private var currentLanguageOption: LanguageOption? = null


    fun subscribeLanguageChanges(activity: Activity) {
        subscribedBroadcastReceivers.getOrPut(activity.hashCode()) { LanguageChangeReceiver() }
            .bindActivity(activity)
    }

    fun unsubscribeLanguageChanges(activity: Activity) {
        subscribedBroadcastReceivers[activity.hashCode()]?.unbindActivity(activity)
    }

    fun getSystemCurrentLocale(): Locale {
        return Resources.getSystem().configuration.locale
    }

    fun getFollowingSystemOption(): LanguageOption {
        return followingSystem()
    }

    @Synchronized
    private fun getCachedLanguageOption(): LanguageOption {
        val cachedLocale = AppPreferences.getDefault()
            .getSerializable<LanguageOption>(PREF_SELECTED_LOCALE)
        if (cachedLocale != null) {
            for (locale in supportedLanguages) {
                if (locale.id == cachedLocale.id) {
                    return locale
                }
            }
        }
        return getFollowingSystemOption()
    }

    @Synchronized
    fun getCurrentLanguageOption(): LanguageOption? {
        if (currentLanguageOption == null) {
            currentLanguageOption = getCachedLanguageOption()
        }
        return currentLanguageOption
    }

    @Synchronized
    fun setLanguageFollowingSystem() {
        setLanguage(followingSystem())
    }

    @Synchronized
    fun setLanguageWithBroadcast(option: LanguageOption?) {
        setLanguage(option)
        val broadcastIntent = Intent()
        broadcastIntent.setAction(LanguageChangeReceiver.LANGUAGE_CHANGED_ACTION)
        globalContext.sendBroadcast(broadcastIntent)
    }

    @Synchronized
    fun setLanguageWithRestart(option: LanguageOption?) {
        setLanguage(option)
        Single.just(0)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .doOnSuccess { o: Int? ->
            Process.killProcess(Process.myPid())
            System.exit(0)
        }
        .subscribe()
    }

    @Synchronized
    fun setLanguage(option: LanguageOption?) {
        AppPreferences.getDefault().put(PREF_SELECTED_LOCALE, option)
        currentLanguageOption = option
        val res = globalContext.resources
        changeResourcesConfig(res)
    }

    @Synchronized
    fun updateLanguageSetting(context: Context) {
        changeResourcesConfig(context.resources)
    }

    @Synchronized
    fun doOnContextGetResources(res: Resources) {
        val conf = res.configuration
        conf.locale = getCurrentLanguageOption()!!.locale
    }

    @Synchronized
    private fun changeResourcesConfig(res: Resources) {
        val dm = res.displayMetrics
        val conf = res.configuration
        conf.locale = getCurrentLanguageOption()!!.locale
        res.updateConfiguration(conf, dm)
    }
}
